libazureinit/src/util.rs

```diff
@@ -0,0 +1,19 @@
+use futures::{pin_mut, Future};
+use std::time::Duration;
+use tokio::time::{sleep, timeout};
+
+// Run delayer function every delay_sec seconds, for max_duration seconds at maximum.
+pub async fn retry_loop<T>(
+    delayer: impl Future<Output = T>,
+    delay_sec: Duration,
+    max_duration: Duration,
+) -> T {
+    pin_mut!(delayer);
+
+    if let Ok(result) = timeout(max_duration, &mut delayer).await {
+        result
+    } else {
+        sleep(delay_sec).await;
+        delayer.await
+    }
```

Comment:
I think what this does is await the given future once, and if it times out sleeps for `delay_sec` before awaiting the future again without a timeout.